ember-server/src/main/scala/org/http4s/ember/server/internal/WebSocketHelpers.scala

```diff
@@ -0,0 +1,278 @@
+/*
+ * Copyright 2019 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.ember.server.internal
+
+import cats.MonadThrow
+import cats.data.NonEmptyList
+import cats.effect.{Async, Concurrent, Ref}
+import cats.syntax.all._
+import fs2.{Chunk, Pipe, Pull, Stream}
+import fs2.io.net._
+import org.http4s.syntax.all._
+import org.http4s._
+import org.http4s.websocket.{FrameTranscoder, WebSocketContext}
+import org.http4s.headers._
+import org.http4s.ember.core.Read
+import org.http4s.ember.core.Util.timeoutMaybe
+import org.http4s.headers.Connection
+import org.http4s.websocket.{Rfc6455, WebSocketCombinedPipe, WebSocketFrame, WebSocketSeparatePipe}
+import org.typelevel.ci._
+import scodec.bits.ByteVector
+
+import scala.concurrent.duration.Duration
+import java.security.MessageDigest
+import java.nio.charset.StandardCharsets
+import java.nio.ByteBuffer
+import org.typelevel.log4cats.Logger
+import fs2.concurrent.SignallingRef
+
+object WebSocketHelpers {
+
+  private[this] val supportedWebSocketVersion = 13L
+
+  private[this] val upgradeCi = ci"upgrade"
+  private[this] val webSocketProtocol = Protocol(ci"websocket", None)
+  private[this] val connectionUpgrade = Connection(NonEmptyList.of(upgradeCi))
+  private[this] val upgradeWebSocket = Upgrade(webSocketProtocol)
+
+  // TODO followup: use websocketcontext responses for error modes
+  def upgrade[F[_]](
+      socket: Socket[F],
+      req: Request[F],
+      ctx: WebSocketContext[F],
+      buffer: Array[Byte],
+      receiveBufferSize: Int,
+      idleTimeout: Duration,
+      onWriteFailure: (Option[Request[F]], Response[F], Throwable) => F[Unit],
+      errorHandler: Throwable => F[Response[F]],
+      logger: Logger[F])(implicit F: Async[F]): F[Unit] = {
```

Comment:
This moves from `Concurrent` to `Async` because there's a `Sync` buried within, and I didn't think too hard about how to avoid that.  It might be possible to walk that back, but we'd probably have to do assure ourselves some side effect is not observable.